### PR TITLE
[NTOS:MM] Make Working Set Lists usable

### PR DESCRIPTION
**Pull request by zefklop**
## Purpose

Split #3300 into something which can be hopefully reviewed.

Implement initializing, adding & removing entries into working set list.
Not used yet, but can be tested in #3300.
With some C++ inside the kernel :open_mouth: 

## Proposed changes

At process initialisation time, properly initialize the working set list.
Implement adding entries to it (private pages only for now).
Implement removing entries from it (private pages only).
Implement growing/shrinking the WSLE array.
Implement trimming it.

Ready to be reviewed.

### CODE REVIEW

File: ntoskrnl/mm/ARM3/wslist.cpp [hunk 1]
```diff
@@ -0,0 +1,503 @@
...
+class MMWSL_impl : public MMWSL
+{
+private:
+    MMPTE GetPteTemplate()
+    {
+        return (this == MmSystemCacheWorkingSetList) ? ValidKernelPte : ValidKernelPteLocal;
+    }
+
+    ULONG GetNextPageColor()
+    {
```
Comment (HBelusca) [inline ntoskrnl/mm/ARM3/wslist.cpp:35]: Is that an april fools joke?

**Review by HBelusca — CHANGES_REQUESTED** [review]

File: ntoskrnl/mm/ARM3/wslist.cpp (hunk 1 above)
Comment (zefklop) [reply ntoskrnl/mm/ARM3/wslist.cpp:35]: Why would it be ?

Comment (HBelusca) [reply ntoskrnl/mm/ARM3/wslist.cpp:35]: > Why would it be ?

You are developing an **NT kernel**, not HaikuOS or your average game/app/whatever.

Comment (zefklop) [reply ntoskrnl/mm/ARM3/wslist.cpp:35]: And ?

**Comment by Extravert-ir** [comment]
I'm against using C++ in the kernel in such form. It looks like a hack, and makes this code a temporary stub and a candidate for a rewrite in future. 

If we want to use C++ in the kernel, it has to be a more engineered solution - cpp-friendly internal APIs and headers, maybe even different code organization. (Reminds me a Monstera idea from @bragin, see 4ea90432a9199251b7a8093fd355131ee5a5c819)

**Comment by zefklop** [comment]
> I'm against using C++ in the kernel in such form. It looks like a hack, and makes this code a temporary stub and a candidate for a rewrite in future.
> 
> If we want to use C++ in the kernel, it has to be a more engineered solution - cpp-friendly internal APIs and headers, maybe even different code organization. (Reminds me a Monstera idea from @bragin, see [4ea9043](https://github.com/reactos/reactos/commit/4ea90432a9199251b7a8093fd355131ee5a5c819))

Alright, I removed the MMWSL_Impl class, which didn't really add anything anyway.

File: ntoskrnl/mm/ARM3/miarm.h [hunk 2]
```diff
@@ -1291,22 +1341,22 @@ MiUnlockWorkingSet(IN PETHREAD Thread,
...
         Thread->OwnsSystemWorkingSetExclusive = FALSE;
     }
     else if (WorkingSet->Flags.SessionSpace)
     {
         /* Release the session working set */
-        ASSERT((Thread->OwnsSessionWorkingSetExclusive == TRUE) ||
-               (Thread->OwnsSessionWorkingSetShared == TRUE));
+        ASSERT((Thread->OwnsSessionWorkingSetExclusive == TRUE) &&
+               (Thread->OwnsSessionWorkingSetShared == FALSE));
         Thread->OwnsSessionWorkingSetExclusive = 0;
```
Comment (HBelusca) [inline ntoskrnl/mm/ARM3/miarm.h:1353]:
```suggestion
        Thread->OwnsSessionWorkingSetExclusive = FALSE;
```

File: ntoskrnl/mm/ARM3/miarm.h [hunk 3]
```diff
@@ -1317,6 +1367,84 @@ MiUnlockWorkingSet(IN PETHREAD Thread,
...
+        ASSERT((Thread->OwnsSystemWorkingSetExclusive == FALSE) &&
+               (Thread->OwnsSystemWorkingSetShared == TRUE));
+        Thread->OwnsSystemWorkingSetShared = FALSE;
+    }
+    else if (WorkingSet->Flags.SessionSpace)
+    {
+        /* Release the session working set */
+        ASSERT((Thread->OwnsSessionWorkingSetExclusive == FALSE) &&
+               (Thread->OwnsSessionWorkingSetShared == TRUE));
+        Thread->OwnsSessionWorkingSetShared = 0;
```
Comment (HBelusca) [inline ntoskrnl/mm/ARM3/miarm.h:1391]:
```suggestion
        Thread->OwnsSessionWorkingSetShared = FALSE;
```

File: ntoskrnl/mm/ARM3/miarm.h [hunk 4]
```diff
@@ -1317,6 +1367,84 @@ MiUnlockWorkingSet(IN PETHREAD Thread,
     KeLeaveGuardedRegion();
 }
 
+FORCEINLINE
+VOID
+MiUnlockWorkingSetShared(IN PETHREAD Thread,
+                   IN PMMSUPPORT WorkingSet)
```
Comment (HBelusca) [inline ntoskrnl/mm/ARM3/miarm.h:1373]:
```suggestion
MiUnlockWorkingSetShared(
    _In_ PETHREAD Thread,
    _In_ PMMSUPPORT WorkingSet)
```

File: ntoskrnl/mm/ARM3/procsup.c [hunk 5]
```diff
@@ -1229,39 +1208,12 @@ MmCreateProcessAddressSpace(IN ULONG MinWs,
...
 
     /* Set the base directory pointers */
-    Process->WorkingSetPage = WsListIndex;
     DirectoryTableBase[0] = PdeIndex << PAGE_SHIFT;
     DirectoryTableBase[1] = HyperIndex << PAGE_SHIFT;
 
     /* Make sure we don't already have a page directory setup */
     ASSERT(Process->Pcb.DirectoryTableBase[0] == 0);
 
-    /* Get a PTE to map hyperspace */
```
Comment (HBelusca) [inline ntoskrnl/mm/ARM3/procsup.c:1239]: Isn't that part of the code covered by the other PR 3572?

Comment (zefklop) [reply ntoskrnl/mm/ARM3/procsup.c:1239]: Yes. One of the two will eventually need to be rebased. But this is needed to have both branches in a working state.

File: ntoskrnl/mm/ARM3/procsup.c [hunk 6]
```diff
@@ -997,7 +966,13 @@ MmInitializeProcessAddressSpace(IN PEPROCESS Process,
     MiInitializePfn(PageFrameNumber, PointerPte, TRUE);
 
     /* Do the same for hyperspace */
-    PointerPde = MiAddressToPde((PVOID)HYPER_SPACE);
```
Comment (Extravert-ir) [inline ntoskrnl/mm/ARM3/procsup.c:1000]: Lol, how did it work before for 64 bits?

File: ntoskrnl/mm/ARM3/wslist.cpp [hunk 7]
```diff
@@ -0,0 +1,484 @@
...
+    /* From now on, every added page can be trimmed at any time */
+    WsList->FirstDynamic = WsList->LastEntry;
+
+    /* We can add this to our list */
+    ExInterlockedInsertTailList(&MmWorkingSetExpansionHead, &WorkingSet->WorkingSetExpansionLinks, &MmExpansionLock);
+}
+
+VOID
+NTAPI
+MmWorkingSetManager(VOID)
```
Comment (Extravert-ir) [inline ntoskrnl/mm/ARM3/wslist.cpp:404]: You don't want to enable it in this PR?

File: ntoskrnl/mm/ARM3/wslist.cpp [hunk 8]
```diff
@@ -0,0 +1,484 @@
...
+_Requires_exclusive_lock_held_(WorkingSet->WorkingSetMutex)
+VOID
+NTAPI
+MiInitializeWorkingSetList(_Inout_ PMMSUPPORT WorkingSet)
+{
+    PMMWSL WsList = WorkingSet->VmWorkingSetList;
+
+    /* Initialize some fields */
+    WsList->FirstFree = ULONG_MAX;
+    WsList->Wsle = reinterpret_cast<PMMWSLE>(WsList + 1);
```
Comment (Extravert-ir) [inline ntoskrnl/mm/ARM3/wslist.cpp:368]: This member is a pointer to the last entry?

File: ntoskrnl/mm/ARM3/wslist.cpp [hunk 9]
```diff
@@ -0,0 +1,484 @@
...
+    WsList->LastEntry = 0;
+    /* The first page is already allocated */
+    WsList->LastInitializedWsle = (PAGE_SIZE - sizeof(*WsList)) / sizeof(MMWSLE);
+
+    /* Insert the address we already know: our PDE base and the Working Set List */
+    if (MI_IS_PROCESS_WORKING_SET(WorkingSet))
+    {
+        ASSERT(WorkingSet->VmWorkingSetList == MmWorkingSetList);
+#if _MI_PAGING_LEVELS == 4
+        MiInsertInWorkingSetList(WorkingSet, (PVOID)PXE_BASE, 0U);
```
Comment (Extravert-ir) [inline ntoskrnl/mm/ARM3/wslist.cpp:378]: The `Protection` parameter is equal to some flag, isn't it?

**Comment by Extravert-ir** [comment]
An assertion comes up in 64bit mode (1st stage setup):
```
Assertion ntoskrnl\mm\ARM3\pfnlist.c(999): Pfn1->u3.e2.ReferenceCount == 0
 # Child-SP          RetAddr           Call Site
00 fffff880`04ff3fb0 fffff800`004fd76b nt!MiInitializePfn+0x17e [ntoskrnl\mm\ARM3\pfnlist.c @ 999] 
01 fffff880`04ff4040 fffff800`0055a404 nt!MmInitializeProcessAddressSpace+0x3fb [ntoskrnl\mm\ARM3\procsup.c @ 994] 
02 fffff880`04ff4200 fffff800`005584e3 nt!PspCreateProcess+0x734 [ntoskrnl\ps\process.c @ 629] 
03 fffff880`04ff45b0 fffff800`005583d5 nt!NtCreateProcessEx+0xf3 [ntoskrnl\ps\process.c @ 1385] 
04 fffff880`04ff4620 fffff800`00405433 nt!NtCreateProcess+0xd5 [ntoskrnl\ps\process.c @ 1433] 
05 fffff880`04ff4690 fffff800`005aa241 nt!KiZwSystemService+0x6f
06 fffff880`04ff4870 fffff800`006a9c02 nt!RtlCreateUserProcess+0x261 [sdk\lib\rtl\process.c @ 246] 
07 fffff880`04ff4a10 fffff800`006ab021 nt!ExpLoadInitialProcess+0x602 [ntoskrnl\ex\init.c @ 555] 
08 fffff880`04ff4be0 fffff800`00439988 nt!Phase1InitializationDiscard+0x12e1 [ntoskrnl\ex\init.c @ 1977] 
09 fffff880`04ff4f60 fffff800`00565ee3 nt!Phase1Initialization+0x28 [ntoskrnl\ex\init.c @ 2021] 
0a fffff880`04ff4f90 fffff800`0040266b nt!PspSystemThreadStartup+0x63 [ntoskrnl\ps\thread.c @ 158] 
0b fffff880`04ff4fd0 00000000`00000000 nt!KiThreadStartup+0x3b
```
```
kd> ?? Pfn1->u3.e2.ReferenceCount
unsigned short 1
```

File: ntoskrnl/mm/ARM3/wslist.cpp (hunk 9 above)
Comment (zefklop) [reply ntoskrnl/mm/ARM3/wslist.cpp:378]: To the Mm internal representation of Nt protection bits. MM_READWRITE & the like. See here:
https://git.reactos.org/?p=reactos.git;a=blob;f=ntoskrnl/mm/ARM3/miarm.h;hb=0dedb9b47457e86ca57b93856af3da6cb0138a93#l35

**Comment by zefklop** [comment]
> An assertion comes up in 64bit mode (1st stage setup):
> 
> ```
> Assertion ntoskrnl\mm\ARM3\pfnlist.c(999): Pfn1->u3.e2.ReferenceCount == 0
>  # Child-SP          RetAddr           Call Site
> 00 fffff880`04ff3fb0 fffff800`004fd76b nt!MiInitializePfn+0x17e [ntoskrnl\mm\ARM3\pfnlist.c @ 999] 
> 01 fffff880`04ff4040 fffff800`0055a404 nt!MmInitializeProcessAddressSpace+0x3fb [ntoskrnl\mm\ARM3\procsup.c @ 994] 
> 02 fffff880`04ff4200 fffff800`005584e3 nt!PspCreateProcess+0x734 [ntoskrnl\ps\process.c @ 629] 
> 03 fffff880`04ff45b0 fffff800`005583d5 nt!NtCreateProcessEx+0xf3 [ntoskrnl\ps\process.c @ 1385] 
> 04 fffff880`04ff4620 fffff800`00405433 nt!NtCreateProcess+0xd5 [ntoskrnl\ps\process.c @ 1433] 
> 05 fffff880`04ff4690 fffff800`005aa241 nt!KiZwSystemService+0x6f
> 06 fffff880`04ff4870 fffff800`006a9c02 nt!RtlCreateUserProcess+0x261 [sdk\lib\rtl\process.c @ 246] 
> 07 fffff880`04ff4a10 fffff800`006ab021 nt!ExpLoadInitialProcess+0x602 [ntoskrnl\ex\init.c @ 555] 
> 08 fffff880`04ff4be0 fffff800`00439988 nt!Phase1InitializationDiscard+0x12e1 [ntoskrnl\ex\init.c @ 1977] 
> 09 fffff880`04ff4f60 fffff800`00565ee3 nt!Phase1Initialization+0x28 [ntoskrnl\ex\init.c @ 2021] 
> 0a fffff880`04ff4f90 fffff800`0040266b nt!PspSystemThreadStartup+0x63 [ntoskrnl\ps\thread.c @ 158] 
> 0b fffff880`04ff4fd0 00000000`00000000 nt!KiThreadStartup+0x3b
> ```
> 
> ```
> kd> ?? Pfn1->u3.e2.ReferenceCount
> unsigned short 1
> ```



> An assertion comes up in 64bit mode (1st stage setup):
> 
> ```
> Assertion ntoskrnl\mm\ARM3\pfnlist.c(999): Pfn1->u3.e2.ReferenceCount == 0
>  # Child-SP          RetAddr           Call Site
> 00 fffff880`04ff3fb0 fffff800`004fd76b nt!MiInitializePfn+0x17e [ntoskrnl\mm\ARM3\pfnlist.c @ 999] 
> 01 fffff880`04ff4040 fffff800`0055a404 nt!MmInitializeProcessAddressSpace+0x3fb [ntoskrnl\mm\ARM3\procsup.c @ 994] 
> 02 fffff880`04ff4200 fffff800`005584e3 nt!PspCreateProcess+0x734 [ntoskrnl\ps\process.c @ 629] 
> 03 fffff880`04ff45b0 fffff800`005583d5 nt!NtCreateProcessEx+0xf3 [ntoskrnl\ps\process.c @ 1385] 
> 04 fffff880`04ff4620 fffff800`00405433 nt!NtCreateProcess+0xd5 [ntoskrnl\ps\process.c @ 1433] 
> 05 fffff880`04ff4690 fffff800`005aa241 nt!KiZwSystemService+0x6f
> 06 fffff880`04ff4870 fffff800`006a9c02 nt!RtlCreateUserProcess+0x261 [sdk\lib\rtl\process.c @ 246] 
> 07 fffff880`04ff4a10 fffff800`006ab021 nt!ExpLoadInitialProcess+0x602 [ntoskrnl\ex\init.c @ 555] 
> 08 fffff880`04ff4be0 fffff800`00439988 nt!Phase1InitializationDiscard+0x12e1 [ntoskrnl\ex\init.c @ 1977] 
> 09 fffff880`04ff4f60 fffff800`00565ee3 nt!Phase1Initialization+0x28 [ntoskrnl\ex\init.c @ 2021] 
> 0a fffff880`04ff4f90 fffff800`0040266b nt!PspSystemThreadStartup+0x63 [ntoskrnl\ps\thread.c @ 158] 
> 0b fffff880`04ff4fd0 00000000`00000000 nt!KiThreadStartup+0x3b
> ```
> 
> ```
> kd> ?? Pfn1->u3.e2.ReferenceCount
> unsigned short 1
> ```

Should be fixed now.

File: ntoskrnl/mm/ARM3/wslist.cpp (hunk 8 above)
Comment (zefklop) [reply ntoskrnl/mm/ARM3/wslist.cpp:368]: https://git.reactos.org/?p=reactos.git;a=blob;f=sdk/include/ndk/mmtypes.h;hb=0dedb9b47457e86ca57b93856af3da6cb0138a93#l868

The WsList is at the beginning of a page (in hyper space for processes, at MmSystemCacheWorkingSetList for the system one) and the array of MMWSLE starts right after. Then pages get allocated as the array grows.

File: ntoskrnl/mm/ARM3/wslist.cpp (hunk 7 above)
Comment (zefklop) [reply ntoskrnl/mm/ARM3/wslist.cpp:404]: There's no point, no body adds anything to WS lists yet :-)

File: ntoskrnl/mm/ARM3/procsup.c (hunk 6 above)
Comment (zefklop) [reply ntoskrnl/mm/ARM3/procsup.c:1000]: This actually breaks it. This is a leftover from a bad git rebase.

**Review by Extravert-ir — APPROVED** [review]

**Pull request merged**